orchagent/tunneldecaporch.cpp

```diff
@@ -678,46 +1105,50 @@ bool TunnelDecapOrch::setTunnelAttribute(string field, sai_object_id_t value, sa
  *    @brief sets ips for a particular tunnel. deletes ips that are old and adds new ones
  *
  * Arguments:
- *    @param[in] key - key of the tunnel from APP_DB
- *    @param[in] new_ip_addresses - new destination ip addresses to decap (comes from APP_DB)
- *    @param[in] tunnel_id - the id of the tunnel
+ *    @param[in] tunnel_name - tunnel name from APP_DB
+ *    @param[in] src_ip_str - new source ip address for the decap terms
  *
  * Return Values:
  *    @return true on success and false if there's an error
  */
-bool TunnelDecapOrch::setIpAttribute(string key, IpAddresses new_ip_addresses, sai_object_id_t tunnel_id)
+bool TunnelDecapOrch::setIpAttribute(string tunnel_name, string src_ip_str)
 {
-    TunnelEntry *tunnel_info = &tunnelTable.find(key)->second;
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("Setting source IP for decap terms of tunnel %s to %s", tunnel_name.c_str(), src_ip_str.c_str());
 
-    // make a copy of tunnel_term_info to loop through
-    vector<TunnelTermEntry> tunnel_term_info_copy(tunnel_info->tunnel_term_info);
+    auto tunnel_it = tunnelTable.find(tunnel_name);
+    if (tunnel_it == tunnelTable.end())
+    {
+        SWSS_LOG_INFO("Tunnel %s does not exist", tunnel_name.c_str());
+        return true;
+    }
 
-    tunnel_info->tunnel_term_info.clear();
-    tunnel_info->dst_ip_addrs = new_ip_addresses;
+    TunnelEntry *tunnel_info = &tunnel_it->second;
+    map<swss::IpPrefix, TunnelTermEntry> decap_terms_copy(tunnel_info->tunnel_term_info.begin(), tunnel_info->tunnel_term_info.end());
 
-    // loop through original ips and remove ips not in the new ip_addresses
-    for (auto it = tunnel_term_info_copy.begin(); it != tunnel_term_info_copy.end(); ++it)
+    for (auto it = decap_terms_copy.begin(); it != decap_terms_copy.end(); ++it)
     {
-        TunnelTermEntry tunnel_entry_info = *it;
-        string ip = tunnel_entry_info.dst_ip;
-        if (!new_ip_addresses.contains(ip))
+        TunnelTermEntry &term_entry = it->second;
+        if ((term_entry.term_type == TUNNEL_TERM_TYPE_P2P) || (term_entry.term_type == TUNNEL_TERM_TYPE_MP2MP))
         {
-            if (!removeDecapTunnelTermEntry(tunnel_entry_info.tunnel_term_id, ip))
+            if (!removeDecapTunnelTermEntry(tunnel_name, term_entry.dst_ip))
```

Comment:
How to handle the dst_ip change for the existing P2MP tunnel term?

Reply:
We only cover source IP change, and currently, we use `MP2P` without source IP, so let's keep current usage for now and correct it based on the SAI API in the future.